cdap-common/src/main/java/co/cask/cdap/common/dataset/DatasetClassRewriter.java

```diff
@@ -0,0 +1,251 @@
+      interfaceClass = Modifier.isInterface(access);
+    }
+
+    @Override
+    public FieldVisitor visitField(int access, String name, String desc, String signature, Object value) {
+      fields.add(name);
+      return super.visitField(access, name, desc, signature, value);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(int access, final String name,
+                                     String desc, String signature, String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, name, desc, signature, exceptions);
+
+      // No need to if the class is an interface, the method is static or the method should be ignored
+      if (interfaceClass || Modifier.isStatic(access) || IGNORE_METHODS.contains(new Method(name, desc))) {
+        return mv;
+      }
+
+      if (datasetRuntimeContextField == null) {
+        // Generate the DatasetRuntimeContextField.
+        // visitMethod is always called after all visitField calls, hence generate the extra field here
+        datasetRuntimeContextField = generateFieldName("_datasetRuntimeContext", fields);
+        super.visitField(Modifier.PRIVATE, datasetRuntimeContextField,
+                         DATASET_RUNTIME_CONTEXT_TYPE.getDescriptor(), null, null);
+      }
+
+      return new FinallyAdapter(Opcodes.ASM5, mv, access, name, desc) {
+
+        private boolean hasRead;
+        private boolean hasWrite;
+        private boolean hasNoAccess;
+
+        @Override
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+          if (visible) {
+            String annotation = Type.getType(desc).getClassName();
+            if (ReadOnly.class.getName().equals(annotation)) {
+              hasRead = true;
+            } else if (WriteOnly.class.getName().equals(annotation)) {
+              hasWrite = true;
+            } else if (ReadWrite.class.getName().equals(annotation)) {
+              hasRead = true;
+              hasWrite = true;
+            } else if (NoAccess.class.getName().equals(annotation)) {
+              hasNoAccess = true;
+            }
+          }
+          return super.visitAnnotation(desc, visible);
+        }
+
+        @Override
+        protected void onMethodEnter() {
+          boolean isConstructor = "<init>".equals(name);
+          if (isConstructor) {
+            // this._datasetRuntimeContext = DatasetRuntimeContext.getContext();
+            loadThis();
+            invokeStatic(DATASET_RUNTIME_CONTEXT_TYPE,
+                         new Method("getContext", Type.getMethodDescriptor(DATASET_RUNTIME_CONTEXT_TYPE)));
+            putField(datasetType, datasetRuntimeContextField, DATASET_RUNTIME_CONTEXT_TYPE);
+          }
+
+          Type methodAnnotationType = getMethodAnnotationType(hasRead, hasWrite, hasNoAccess);
+          Type defaultAnnotationType = isConstructor ? Type.getType(NoAccess.class) : Type.getType(ReadWrite.class);
```

Comment:
would it make sense to have an UnknownAccess as the default if neither of the other ones is present? It would be enforced as if it is a @ReadWrite, but for lineage or audit purpose, it could be logged as Unkown access.